File: php/src/VedaRetryPolicy.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Retry policy with configurable backoff strategies.
 */
class VedaRetryPolicy
{
    private int $maxRetries;
    private float $baseDelayMs;
    private float $maxDelayMs;
    private float $multiplier;
    private string $strategy; // 'fixed', 'linear', 'exponential', 'jitter'

    /** @var list<class-string<\Throwable>> */
    private array $retryableExceptions;

    /**
     * @param list<class-string<\Throwable>> $retryableExceptions
     */
    public function __construct(
        int $maxRetries = 3,
        float $baseDelayMs = 100.0,
        float $maxDelayMs = 10000.0,
        float $multiplier = 2.0,
        string $strategy = 'exponential',
        array $retryableExceptions = [ConnectionException::class, TimeoutException::class]
    ) {
        $this->maxRetries            = max(0, $maxRetries);
        $this->baseDelayMs           = max(1.0, $baseDelayMs);
        $this->maxDelayMs            = max($this->baseDelayMs, $maxDelayMs);
        $this->multiplier            = max(1.0, $multiplier);
        $this->strategy              = $strategy;
        $this->retryableExceptions   = $retryableExceptions;
    }

    /**
     * Execute a callback with retry logic.
     *
     * @template T
     * @param callable(): T $operation
     * @return T
     */
    public function execute(callable $operation): mixed
    {
        $lastException = null;

        for ($attempt = 0; $attempt <= $this->maxRetries; $attempt++) {
            try {
                return $operation();
            } catch (\Throwable $e) {
                $lastException = $e;

                if (!$this->isRetryable($e) || $attempt >= $this->maxRetries) {
                    throw $e;
                }

                $delay = $this->calculateDelay($attempt);
                usleep((int) ($delay * 1000));
            }
        }

        throw $lastException ?? new VedaException('Retry exhausted');
    }

    /**
     * Check if an exception is retryable.
     */
    public function isRetryable(\Throwable $e): bool
    {
        foreach ($this->retryableExceptions as $class) {
            if ($e instanceof $class) {
                return true;
            }
        }
        return false;
    }

    /**
     * Calculate the delay for a given attempt number.
     */
    public function calculateDelay(int $attempt): float
    {
        $delay = match ($this->strategy) {
            'fixed'        => $this->baseDelayMs,
            'linear'       => $this->baseDelayMs * ($attempt + 1),
            'exponential'  => $this->baseDelayMs * ($this->multiplier ** $attempt),
            'jitter'       => $this->baseDelayMs * ($this->multiplier ** $attempt) * (mt_rand() / mt_getrandmax()),
            default        => $this->baseDelayMs,
        };

        return min($delay, $this->maxDelayMs);
    }

    /**
     * Create a default retry policy.
     */
    public static function default(): self
    {
        return new self();
    }

    /**
     * Create an aggressive retry policy.
     */
    public static function aggressive(): self
    {
        return new self(maxRetries: 5, baseDelayMs: 50, strategy: 'jitter');
    }

    /**
     * Create a conservative retry policy.
     */
    public static function conservative(): self
    {
        return new self(maxRetries: 2, baseDelayMs: 500, multiplier: 2.0, strategy: 'exponential');
    }

    public function getMaxRetries(): int
    {
        return $this->maxRetries;
    }

    public function getBaseDelayMs(): float
    {
        return $this->baseDelayMs;
    }

    public function getMaxDelayMs(): float
    {
        return $this->maxDelayMs;
    }
}

File: dotnet/VedaRetryPolicy.cs
using System;

namespace VedaDB;

/// <summary>
/// Configuration for retry behavior with exponential backoff.
/// </summary>
public class VedaRetryConfig
{
    /// <summary>
    /// Maximum number of retry attempts. Default is 3.
    /// </summary>
    public int MaxRetries { get; set; } = 3;

    /// <summary>
    /// Initial delay between retries in milliseconds. Default is 1000.
    /// </summary>
    public int InitialDelayMs { get; set; } = 1000;

    /// <summary>
    /// Backoff multiplier. Default is 2.0 (exponential).
    /// </summary>
    public double BackoffMultiplier { get; set; } = 2.0;

    /// <summary>
    /// Maximum delay between retries in milliseconds. Default is 30000.
    /// </summary>
    public int MaxDelayMs { get; set; } = 30000;

    /// <summary>
    /// Optional jitter factor (0.0 to 1.0) to randomize delays. Default is 0.1.
    /// </summary>
    public double JitterFactor { get; set; } = 0.1;

    /// <summary>
    /// Exception types that should trigger a retry.
    /// </summary>
    public List<Type> RetryableExceptions { get; set; } = new()
    {
        typeof(VedaConnectionException),
        typeof(TimeoutException),
        typeof(IOException)
    };

    /// <summary>
    /// Custom predicate to determine if an exception is retryable.
    /// </summary>
    public Func<Exception, bool>? IsRetryable { get; set; }
}

/// <summary>
/// Retry policy with exponential backoff for VedaDB operations.
/// </summary>
public class VedaRetryPolicy
{
    private readonly VedaRetryConfig _config;
    private readonly Random _random = new();

    /// <summary>
    /// Create a retry policy with default configuration.
    /// </summary>
    public VedaRetryPolicy() : this(new VedaRetryConfig()) { }

    /// <summary>
    /// Create a retry policy with the specified configuration.
    /// </summary>
    public VedaRetryPolicy(VedaRetryConfig config)
    {
        _config = config ?? throw new ArgumentNullException(nameof(config));
    }

    /// <summary>
    /// Execute an async operation with retry.
    /// </summary>
    public async Task<T> ExecuteAsync<T>(Func<Task<T>> operation, CancellationToken ct = default)
    {
        Exception? lastException = null;

        for (int attempt = 0; attempt <= _config.MaxRetries; attempt++)
        {
            try
            {
                VedaMetrics.TimerStart("vedadb_retry_operation");
                var result = await operation();
                VedaMetrics.TimerStop("vedadb_retry_operation");

                if (attempt > 0)
                    VedaMetrics.Increment("vedadb_retry_success_after_retry", 1, new() { { "attempts", attempt.ToString() } });

                return result;
            }
            catch (Exception ex)
            {
                lastException = ex;
                VedaMetrics.Increment("vedadb_retry_attempt", 1, new() { { "attempt", attempt.ToString() } });

                if (attempt == _config.MaxRetries || !IsRetryable(ex))
                    break;

                var delay = CalculateDelay(attempt);
                await Task.Delay(delay, ct);
            }
        }

        VedaMetrics.Increment("vedadb_retry_exhausted");
        throw new VedaRetryExhaustedException(
            $"Operation failed after {_config.MaxRetries + 1} attempts", lastException!);
    }

    /// <summary>
    /// Execute an async operation without a return value.
    /// </summary>
    public async Task ExecuteAsync(Func<Task> operation, CancellationToken ct = default)
    {
        await ExecuteAsync(async () => { await operation(); return true; }, ct);
    }

    /// <summary>
    /// Execute a sync operation with retry.
    /// </summary>
    public T Execute<T>(Func<T> operation)
    {
        Exception? lastException = null;

        for (int attempt = 0; attempt <= _config.MaxRetries; attempt++)
        {
            try
            {
                return operation();
            }
            catch (Exception ex)
            {
                lastException = ex;
                if (attempt == _config.MaxRetries || !IsRetryable(ex))
                    break;

                var delay = CalculateDelay(attempt);
                Thread.Sleep(delay);
            }
        }

        throw new VedaRetryExhaustedException(
            $"Operation failed after {_config.MaxRetries + 1} attempts", lastException!);
    }

    /// <summary>
    /// Calculate the delay for a given retry attempt.
    /// </summary>
    public TimeSpan CalculateDelay(int attempt)
    {
        var baseDelay = _config.InitialDelayMs * Math.Pow(_config.BackoffMultiplier, attempt);
        var cappedDelay = Math.Min(baseDelay, _config.MaxDelayMs);

        if (_config.JitterFactor > 0)
        {
            var jitter = cappedDelay * _config.JitterFactor * (_random.NextDouble() * 2 - 1);
            cappedDelay = Math.Max(1, cappedDelay + jitter);
        }

        return TimeSpan.FromMilliseconds(cappedDelay);
    }

    /// <summary>
    /// Check if an exception is retryable.
    /// </summary>
    public bool IsRetryable(Exception ex)
    {
        if (_config.IsRetryable != null)
            return _config.IsRetryable(ex);

        return _config.RetryableExceptions.Any(t => t.IsInstanceOfType(ex));
    }

    /// <summary>
    /// Create a default retry policy.
    /// </summary>
    public static VedaRetryPolicy Default => new();

    /// <summary>
    /// Create an aggressive retry policy with more retries and shorter delays.
    /// </summary>
    public static VedaRetryPolicy Aggressive => new(new VedaRetryConfig
    {
        MaxRetries = 5,
        InitialDelayMs = 500,
        BackoffMultiplier = 1.5
    });

    /// <summary>
    /// Create a conservative retry policy with fewer retries and longer delays.
    /// </summary>
    public static VedaRetryPolicy Conservative => new(new VedaRetryConfig
    {
        MaxRetries = 2,
        InitialDelayMs = 2000,
        BackoffMultiplier = 2.0,
        MaxDelayMs = 60000
    });
}
